### PR TITLE
Infrastructure app provision

### PR DESCRIPTION
**Pull request by freva**

### CODE REVIEW

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            }
+        }
+    }
+
+    /**
+     * Returns the version that the given node type should be provisioned to. This is
+     * the version returned by {@link InfrastructureVersions#getTargetVersionFor} unless a provisioning is:
+     * <ul>
+     *   <li>not possible: no nodes of given type in legal state in node-repo</li>
+     *   <li>redudant: all nodes that can be provisioned already have the right wanted Vespa version</li>
```
Comment (mpolden) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java]: "redudant" -> "redundant"

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+     * </ul>
+     */
+    Optional<Version> getVersionToProvision(NodeType nodeType) {
+        Optional<Version> wantedWantedVersion = infrastructureVersions.getTargetVersionFor(nodeType);
```
Comment (mpolden) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java]: `wantedWantedVersion` -> `targetVersion`

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+    /**
+     * Returns the version that the given node type should be provisioned to. This is
+     * the version returned by {@link InfrastructureVersions#getTargetVersionFor} unless a provisioning is:
+     * <ul>
+     *   <li>not possible: no nodes of given type in legal state in node-repo</li>
+     *   <li>redudant: all nodes that can be provisioned already have the right wanted Vespa version</li>
+     * </ul>
+     */
+    Optional<Version> getVersionToProvision(NodeType nodeType) {
```
Comment (mpolden) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java]: Consider naming `targetVersion`.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureVersions.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import java.util.logging.Logger;
+
+/**
+ * Multithread safe class to see and set target versions for infrastructure node types.
+ * {@link InfrastructureProvisioner} maintainer will then allocate the nodes of given node type
+ * with a wantedVersionVersion equal to the given target version.
+ *
+ * @author freva
+ */
+public class InfrastructureVersions {
```
Comment (mpolden) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureVersions.java]: Please add javadoc providing a short description of this.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/testutils/MockProvisioner.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Comment (mpolden) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/testutils/MockProvisioner.java]: Should be updated with current year.

**Review by mpolden — CHANGES_REQUESTED** [review]
Looks great overall. Please have a look at my comments.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureVersions.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    public void setTargetVersion(NodeType nodeType, Version version, boolean force) {
+        if (nodeType != NodeType.config && nodeType != NodeType.confighost && nodeType != NodeType.proxyhost) {
+            throw new IllegalArgumentException("Cannot set version for type " + nodeType);
+        }
+
+        try (Lock lock = db.lockInfrastructureVersions()) {
+            Map<NodeType, Version> infrastructureVersions = db.readInfrastructureVersions();
+            Optional<Version> currentWantedVersion = Optional.ofNullable(infrastructureVersions.get(nodeType));
```
[the remark this answers is not in the document]
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureVersions.java]: Renamed to `currentTargetVersion` and the argument to `newTargetVersion`

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureVersions.java (hunk 4 above)
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureVersions.java]: It already has a short javadoc description?

Comment (mpolden) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureVersions.java]: Sorry, my mistake. 🙂

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+     * </li>
+     */
+    Optional<Version> getVersionToProvision(NodeType nodeType) {
+        Optional<Version> wantedWantedVersion = infrastructureVersions.getTargetVersionFor(nodeType);
+        if (!wantedWantedVersion.isPresent()) {
+            logger.log(LogLevel.DEBUG, "Skipping provision of " + nodeType + ": No target version set");
+            return Optional.empty();
+        }
+
+        List<Optional<Version>> currentWantedVersions = nodeRepository().getNodes(nodeType,
```
[the remark this answers is not in the document]
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java]: Agreed. Looked over the code again, should be consistent now.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/**
+ * @author freva
+ */
+public class InfrastructureProvisioner extends Maintainer {
```
[the remark this answers is not in the document]
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java]: Added.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureVersions.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+            // Trying to set the version to the current version, skip
+            if (currentWantedVersion.equals(Optional.of(version))) return;
+
+            // If we don't force the set, we must set the new version to higher than the already set version
+            if (!force) {
+                if (currentWantedVersion.map(curVersion -> curVersion.compareTo(version) > 0).orElse(false))
```
[the remark this answers is not in the document]
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureVersions.java]: Nice, didn't notice those methods. Fixed.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+    /**
+     * Returns the version that the given node type should be provisioned to. This is
+     * the version returned by {@link InfrastructureVersions#getTargetVersionFor} unless a provisioning is:
+     * <li>
+     *   <ul>not possible: no nodes of given type in legal state in node-repo</ul>
+     *   <ul>redudant: all nodes that can be provisioned already have the right wanted Vespa version</ul>
+     * </li>
+     */
+    Optional<Version> getVersionToProvision(NodeType nodeType) {
```
[the remark this answers is not in the document]
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java]: Fixed.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/v2/UpgradeResponse.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+import com.yahoo.slime.JsonFormat;
+import com.yahoo.slime.Slime;
+import com.yahoo.vespa.hosted.provision.maintenance.InfrastructureVersions;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Comparator;
+import java.util.Map;
+
+/** A response containing infrastructure versions */
```
[the remark this answers is not in the document]
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/v2/UpgradeResponse.java]: Fixed.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureVersions.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.logging.Logger;
+
+/**
+ * Multithread safe class to see and set target versions for infrastructure node types.
+ * {@link InfrastructureProvisioner} maintainer will then allocate the nodes of given node type
+ * with a wantedVersionVersion equal to the given target version.
```
[the remark this answers is not in the document]
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureVersions.java]: Fixed.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java (hunk 1 above)
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java]: Fixed.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/testutils/MockProvisioner.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+import com.yahoo.config.provision.HostSpec;
+import com.yahoo.config.provision.ProvisionLogger;
+import com.yahoo.config.provision.Provisioner;
+import com.yahoo.transaction.NestedTransaction;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+public class MockProvisioner implements Provisioner {
```
[the remark this answers is not in the document]
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/testutils/MockProvisioner.java]: Fixed.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java (hunk 2 above)
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java]: Fixed.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java (hunk 3 above)
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java]: Fixed.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/testutils/MockProvisioner.java (hunk 5 above)
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/testutils/MockProvisioner.java]: Fixed.

**Review by mpolden — APPROVED** [review]

**Pull request merged**